cloud-controller-manager/do/firewall_controller.go

```diff
@@ -0,0 +1,359 @@
+// meaning it did not expect to see any more of its firewalls created or deleted. This function is not meant to be
+// invoked concurrently with the same key.
+func (fc *Controller) syncService(key string) error {
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).Infof("Finished syncing service %q (%v)", key, time.Since(startTime))
+	}()
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	// service holds the latest service info from apiserver
+	service, err := fc.serviceLister.Services(namespace).Get(name)
+	switch {
+	case errors.IsNotFound(err):
+		// cleanup firewall and service cache
+		fc.svcCache.delete(key)
+		fc.fwCache.delete(key)
+	case err != nil:
+		runtime.HandleError(fmt.Errorf("Unable to retrieve service %v from store: %v", key, err))
+	default:
+		err = fc.processServiceCreateOrUpdate(service, key)
+	}
+
+	return err
+}
+
+// processServiceCreateOrUpdate modifies (or not) the firewall for the incoming service accordingly.
+// Returns an error if processing the service update failed.
+func (fc *Controller) processServiceCreateOrUpdate(service *v1.Service, key string) error {
+	cachedService := fc.svcCache.getOrCreate(key)
+	if cachedService.state != nil && cachedService.state.UID != service.UID {
+		// This happens only when a service is deleted and re-created
+		// in a short period, which is only possible when it doesn't
+		// contain finalizer.
+		if err := fc.processFirewallDelete(key, "", cachedService.state); err != nil {
+			return err
+		}
+	}
+	// Always cache the service, we need the info for firewall modification.
+	cachedService.state = service
+
+	return nil
+}
+
+func (fc *Controller) processFirewallCreateOrUpdate(ctx context.Context, key, workerFirewallName string) error {
+	cachedFirewall := fc.fwCache.getOrCreate(key)
+	// We do not care about other firewalls
+	if cachedFirewall.state.Name != workerFirewallName {
+		return nil
+	}
+
+	// Inbound rules firewall request setup for NodePort range.
+	inboundNodePortRules := &godo.InboundRule{
+		Protocol:  "tcp",
+		PortRange: "30000-32767",
+	}
+	rr := &godo.FirewallRulesRequest{
+		InboundRules: []godo.InboundRule{*inboundNodePortRules},
+	}
+	fr := &godo.FirewallRequest{
+		Name:         workerFirewallName,
+		InboundRules: []godo.InboundRule{*inboundNodePortRules},
+	}
+
+	// Create the firewall if it does not exist, update it if it does.
+	fw, _, _ := fc.firewallService.Get(ctx, cachedFirewall.state.ID)
+	if fw == nil {
+		// Create a new firewall with the worker firewall name and inbound rules for NodePort range.
+		fw, err := fc.create(ctx, workerFirewallName, fr, rr)
+		if err != nil {
+			return fmt.Errorf("Failed to create firewall: %s", err)
+		}
+		cachedFirewall.state = fw
+		return nil
+	} else if fw.Name == workerFirewallName {
+		// Update the existing firewall to include inbound rules for NodePort range.
+		fw, err := fc.update(ctx, fw, rr)
```

Comment:
I think we should only send out an update request after comparing the current firewall rules to the cached state and verifying that there's a diff. We're watching for changes on Services, but not every Service change is going to be relevant for us. The extra check will make sure we won't do unnecessary API calls.

Of course it means we'll have to know the initial state of the firewall. Since we want to have another controller to watch for changes to the firewall resource in the DO API anyway, we should get the data for sure.